Sources/MongoSwift/MongoCollection+Write.swift

```diff
@@ -432,4 +376,21 @@ public struct UpdateResult: Decodable {
 
     /// The number of documents that were upserted.
     public let upsertedCount: Int
+
+    internal init?(from result: BulkWriteResult?) throws {
+        guard let result = result else {
+            return nil
+        }
+        self.matchedCount = result.matchedCount
+        self.modifiedCount = result.modifiedCount
+        self.upsertedCount = result.upsertedCount
+        if result.upsertedCount == 1 {
+            guard let id = result.upsertedIds[0] else {
+                throw RuntimeError.internalError(message: "BulkWriteResult missing _id for upserted document")
+            }
+            self.upsertedId = AnyBSONValue(id)
```

Comment:
so we made `upsertedId` an `AnyBSONValue` so the whole struct would be automatically `Decodable`. however, this is not really consistent with the other result types that have `_id`s in them. they just use `BSONValue`.

we no longer decode to the struct in this file since we construct it from a `BulkWriteResult` instead, so I thought about switching to a `BSONValue` and making the struct non-decodable. but, turns out we are using the fact that it's `Decodable` in the CRUD tests to get the expected result from the JSON files. but we could change things to decode it manually if want, which would make this struct consistent with the rest of the API. 

thoughts?

Reply:
I vote for consistency! Can you explain why it needs to be `AnyBSONValue` for the CRUD tests? This would be a breaking change. IIRC we created `AnyBSONValue` for the `_id` case, since those can be any value. Does it make more sense to standardize on using `AnyBSONValue` in any result struct returning `_id` s?

Reply:
so `AnyBSONValue` exists for the purpose of letting you encode/decode `BSONValue`s without having to know their concrete type. It does the type switching for you in its encode/decode methods. The particular motivation was Stitch having various structs that they stored `_id`s in and they wanted to make those structs `Decodable`. 

so, I *think* we used an `AnyBSONValue` in this struct so we could get free `Decodable` conformance for the `upsertedId` field. The `Decodable` conformance comes into play both when decoding server replies to construct them (we no longer do that as of this PR) and when decoding the CRUD test data into "expected" structs (see [here](https://github.com/mongodb/mongo-swift-driver/blob/088f794b537cdf0e10504e1506d4b16018756661/Tests/MongoSwiftTests/CrudTests.swift#L186)). 

Basically, my guess is that we made this choice for our own convenience, not for the user's. 

While it's a bit more annoying on our end, IMO the result structs should all just use `BSONValue`, not `AnyBSONValue`. Otherwise the user has to do like `let id = result.insertedId.value` instead of `let id = result.insertedId`. 

If we switch to just plain `BSONValue`, we could work around this in the crud tests by writing a manual `Decodable` implementation like this:
```swift
init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    self.matchedCount = try container.decode(Int.self, forKey: .matchedCount)
    self.modifiedCount = try container.decode(Int.self, forKey: .modifiedCount)
    let id = try container.decode(AnyBSONValue.self, forKey: .upsertedId)
    self.upsertedId = id.value
    ...
}
```

Reply:
Thanks for the detailed writeup. I agree, let's make it a `BSONVaue`

Reply:
done